### ISSUE TITLE
extension_loaded breaks MixedAssignment 

### ISSUE THREAD

**Issue by serbancatalin** [vimeo/psalm#6457]
Hello,

https://psalm.dev/r/8cfb4b0115

If you comment if (extension_loaded('invalid')) it will know that $referenceDate is DateTime

**Comment by orklah** [comment]
Can you describe your use case a little?

When Psalm encounters a extension_loaded function, it will check its content to see if the extension is loaded in the environment that runs Psalm.

If it's not, the flag will help skiping part of the analysis to avoid emitting too much error just because the extension is not there.

'invalid' is not detected so Psalm tries its best. There may be a bug lying here because there is no specific code that Psalm should be afraid of, but I'd like to understand what you need this for before diving into the code for possibly hours :p

(Note that if you just need a quick solution, you can do that: https://psalm.dev/r/0fc2dc3293)

**Comment by serbancatalin** [comment]
I am using newrelic extension to name and start/stop a transaction.
This extension is not loaded in the environment that runs Psalm.

Something like 
if (\extension_loaded('newrelic')) {
    newrelic_end_transaction();
    newrelic_ignore_transaction();
    newrelic_start_transaction('test');
    newrelic_name_transaction('test');
    newrelic_background_job(false);
}

Your solution is good.
I can also move this code into another method.

Thank you

**Comment by orklah** [comment]
you should maybe consider loading the extension on Psalm's environment. it may be able to find more bugs when the extension is there.

I'll close this as it seems a pretty edge case and you're satisfied with the quick solution. If I see another similar issue, we'll think about it more.

Thanks for your feedback anyway!

**Comment by nicolashohm** [comment]
Just wanted to let you know that I encountered the same or similar issue.

I have a PHP file as entry point which does check for NewRelic extension at the beginning, but errors after the extension check are ignored by psalm.

Environment: `Target PHP version: 8.1 (inferred from composer.json) Enabled extensions: dom, pdo, redis.` 

Here is a simplified example of my use case:

`entrypoint.php`:
```php
<?php 
require_once __DIR__ . 'vendor/autoload.php';

if (extension_loaded('newrelic')) {
    newrelic_name_transaction('Test');
}

$factory = new Factory();
$factory->createBarz();
```

`Factory.php`
```php
<?php

class Factory {
    public function createFoo() { }
}
```

Psalm runs and reports no errors even though the factory method doesn't exist.

As soon as I extract `extension_loaded('newrelic')` into a method or even just put the extension name in a variable as suggested by @orklah :
```php
$extension = 'newrelic';
if (extension_loaded($extension)) {
```
it works, and psalm reports that `createBarz` does not exist in the Factory.

**Comment by danog** [comment]
Hmm could you try reproducing on psalm.dev? This should have been fixed by https://github.com/vimeo/psalm/pull/10295...

**Comment by orklah** [comment]
There was no release since then so it makes sense the issue is still here on 5.15
